Brokerages/Backtesting/BacktestingBrokerage.cs

```diff
@@ -318,40 +319,21 @@ public virtual void Scan()
                     {
                         //Model:
                         var model = security.FillModel;
-
                         //Based on the order type: refresh its model to get fill price and quantity
                         try
                         {
-                            switch (order.Type)
+                            if (order.Type == OrderType.OptionExercise)
+                            {
+                                var option = (Option)security;
+                                fills = option.OptionExerciseModel.OptionExercise(option, order as OptionExerciseOrder).ToArray();
+                            }
+                            else
                             {
-                                case OrderType.Limit:
-                                    fills = new[] { model.LimitFill(security, order as LimitOrder) };
-                                    break;
-
-                                case OrderType.StopMarket:
-                                    fills = new[] { model.StopMarketFill(security, order as StopMarketOrder) };
-                                    break;
-
-                                case OrderType.Market:
-                                    fills = new[] { model.MarketFill(security, order as MarketOrder) };
-                                    break;
-
-                                case OrderType.StopLimit:
-                                    fills = new[] { model.StopLimitFill(security, order as StopLimitOrder) };
-                                    break;
-
-                                case OrderType.MarketOnOpen:
-                                    fills = new[] { model.MarketOnOpenFill(security, order as MarketOnOpenOrder) };
-                                    break;
-
-                                case OrderType.MarketOnClose:
-                                    fills = new[] { model.MarketOnCloseFill(security, order as MarketOnCloseOrder) };
-                                    break;
-
-                                case OrderType.OptionExercise:
-                                    var option = (Option)security;
-                                    fills = option.OptionExerciseModel.OptionExercise(option, order as OptionExerciseOrder).ToArray();
-                                    break;
+                                var context = new FillModelContext(
+                                    security,
+                                    order,
+                                    Algorithm.SubscriptionManager.SubscriptionDataConfigService);
+                                fills = new[] { model.Fill(context) };
```

Comment:
Sexy! Love it! User defined custom orders here we come!

Hmm... I almost wonder if we should take this opportunity to make this return a context object -- I'm thinking about support multiple fills at the same time step which will become important when we start managing L2 depth data. With L2 depth data we can get _really_ accurate fill models by eating through the book, and each price level you 'eat' would be a separate fill, so it's possible to have multiple fills at the same time step. Figured it may be worthwhile to account for this now while we're doing the semi-breaking changes.